File: runtime/app/crypto.py
# runtime/app/crypto.py — Cifrado de credenciales (compatible con web/lib/crypto.ts)
#
# AES-256-GCM. Formato almacenado: base64(nonce(12) || ciphertext(N) || tag(16)).
# Compatibilidad bit-a-bit con Node crypto en web/lib/crypto.ts.

import base64
import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def _load_key() -> bytes:
    raw = os.getenv("ENCRYPTION_KEY", "")
    if not raw:
        raise RuntimeError("ENCRYPTION_KEY no configurada")
    key = base64.b64decode(raw)
    if len(key) != 32:
        raise RuntimeError(f"ENCRYPTION_KEY debe ser 32 bytes (got {len(key)})")
    return key


def descifrar(token_b64: str) -> str:
    """Descifra un string cifrado por web/lib/crypto.ts. Lanza excepción si falla."""
    if not token_b64:
        raise ValueError("descifrar: token vacío")
    raw = base64.b64decode(token_b64)
    if len(raw) < 28:
        raise ValueError(f"descifrar: payload demasiado corto ({len(raw)} bytes)")
    nonce, body = raw[:12], raw[12:]
    plain = AESGCM(_load_key()).decrypt(nonce, body, None)
    return plain.decode("utf-8")


def cifrar(texto: str) -> str:
    """Cifra un string. Útil para scripts de mantenimiento."""
    nonce = os.urandom(12)
    body = AESGCM(_load_key()).encrypt(nonce, texto.encode("utf-8"), None)
    return base64.b64encode(nonce + body).decode("ascii")

File: web/lib/crypto.ts
// web/lib/crypto.ts — AES-256-GCM compatible con runtime/app/crypto.py
//
// Formato almacenado: base64(nonce(12) || ciphertext(N) || authTag(16))

import crypto from "node:crypto";

function loadKey(): Buffer {
  const raw = process.env.ENCRYPTION_KEY;
  if (!raw) throw new Error("ENCRYPTION_KEY no configurada");
  const key = Buffer.from(raw, "base64");
  if (key.length !== 32) {
    throw new Error(`ENCRYPTION_KEY debe ser 32 bytes (got ${key.length})`);
  }
  return key;
}

export function cifrar(texto: string): string {
  if (!texto) return "";
  const nonce = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv("aes-256-gcm", loadKey(), nonce);
  const ciphertext = Buffer.concat([cipher.update(texto, "utf8"), cipher.final()]);
  const authTag = cipher.getAuthTag();
  return Buffer.concat([nonce, ciphertext, authTag]).toString("base64");
}

export function descifrar(tokenB64: string): string {
  if (!tokenB64) return "";
  const raw = Buffer.from(tokenB64, "base64");
  const nonce = raw.subarray(0, 12);
  const authTag = raw.subarray(raw.length - 16);
  const ciphertext = raw.subarray(12, raw.length - 16);
  const decipher = crypto.createDecipheriv("aes-256-gcm", loadKey(), nonce);
  decipher.setAuthTag(authTag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString("utf8");
}

export function generarKey(): string {
  return crypto.randomBytes(32).toString("base64");
}
